File: server/services/storageQuotas.js
/**
 * Storage Quota Service
 * ======================
 * Tracks and enforces per-organisation storage limits.
 * Storage types: avatars, replays, exports, attachments.
 */

const db = require('../models/db');
const cache = require('./cache');

const STORAGE_CACHE_TTL = 300; // 5 minutes

/**
 * Record storage usage for a resource.
 *
 * @param {string} organisationId
 * @param {string} tenantId
 * @param {string} resourceType - 'avatar', 'replay', 'export', 'attachment'
 * @param {string} resourceId
 * @param {number} sizeBytes
 */
async function recordUsage(organisationId, tenantId, resourceType, resourceId, sizeBytes) {
    await db.query(`
        INSERT INTO storage_usage (organisation_id, tenant_id, resource_type, resource_id, size_bytes)
        VALUES ($1, $2, $3, $4, $5)
        ON CONFLICT (organisation_id, resource_type, resource_id) DO UPDATE SET
            size_bytes = EXCLUDED.size_bytes
    `, [organisationId, tenantId, resourceType, resourceId, sizeBytes]);

    await cache.del(`storage:${organisationId}`);
}

/**
 * Remove a storage record (e.g., when deleting a resource).
 */
async function removeUsage(organisationId, resourceType, resourceId) {
    await db.query(
        'DELETE FROM storage_usage WHERE organisation_id = $1 AND resource_type = $2 AND resource_id = $3',
        [organisationId, resourceType, resourceId]
    );
    await cache.del(`storage:${organisationId}`);
}

/**
 * Get total storage usage in bytes.
 */
async function getTotalUsage(organisationId) {
    const cacheKey = `storage:${organisationId}`;
    const cached = await cache.get(cacheKey);
    if (cached !== null && cached !== undefined) return cached;

    const result = await db.query(
        'SELECT COALESCE(SUM(size_bytes), 0) as total FROM storage_usage WHERE organisation_id = $1',
        [organisationId]
    );

    const total = parseInt(result.rows[0].total, 10);
    await cache.set(cacheKey, total, STORAGE_CACHE_TTL);
    return total;
}

/**
 * Get usage breakdown by resource type.
 */
async function getUsageBreakdown(organisationId) {
    const result = await db.query(`
        SELECT resource_type, COUNT(*) as file_count, SUM(size_bytes) as total_bytes
        FROM storage_usage WHERE organisation_id = $1
        GROUP BY resource_type
    `, [organisationId]);

    const breakdown = {};
    for (const row of result.rows) {
        breakdown[row.resource_type] = {
            fileCount: parseInt(row.file_count, 10),
            totalBytes: parseInt(row.total_bytes, 10),
            totalMb: Math.round(parseInt(row.total_bytes, 10) / (1024 * 1024) * 100) / 100
        };
    }
    return breakdown;
}

/**
 * Check if an organisation has enough storage quota.
 *
 * @param {string} organisationId
 * @param {number} additionalBytes - How many bytes we want to add
 * @param {number} limitMb - Storage limit in MB (from entitlements)
 * @returns {{ allowed: boolean, usedMb: number, limitMb: number, remainingMb: number }}
 */
async function checkQuota(organisationId, additionalBytes, limitMb) {
    if (limitMb === null || limitMb < 0) {
        // Unlimited
        return { allowed: true, usedMb: 0, limitMb: null, remainingMb: null };
    }

    const totalBytes = await getTotalUsage(organisationId);
    const usedMb = totalBytes / (1024 * 1024);
    const remainingMb = limitMb - usedMb;
    const additionalMb = additionalBytes / (1024 * 1024);

    return {
        allowed: usedMb + additionalMb <= limitMb,
        usedMb: Math.round(usedMb * 100) / 100,
        limitMb,
        remainingMb: Math.round(Math.max(0, remainingMb) * 100) / 100
    };
}

module.exports = {
    recordUsage,
    removeUsage,
    getTotalUsage,
    getUsageBreakdown,
    checkQuota
};

File: server-rs/src/services/storage_quotas.rs
use crate::cache::Cache;
use crate::error::AppResult;
use serde::Serialize;

#[derive(Debug, Serialize)]
pub struct StorageBreakdown {
    pub resource_type: String,
    pub total_bytes: i64,
    pub count: i64,
}

pub async fn record_usage(
    db: &sqlx::PgPool,
    organisation_id: &str,
    tenant_id: &str,
    resource_type: &str,
    resource_id: &str,
    size_bytes: i64,
) -> AppResult<()> {
    sqlx::query(
        r#"INSERT INTO storage_usage (organisation_id, tenant_id, resource_type, resource_id, size_bytes)
        VALUES ($1, $2, $3, $4, $5)
        ON CONFLICT (organisation_id, resource_type, resource_id)
        DO UPDATE SET size_bytes = EXCLUDED.size_bytes"#,
    )
    .bind(organisation_id)
    .bind(tenant_id)
    .bind(resource_type)
    .bind(resource_id)
    .bind(size_bytes)
    .execute(db)
    .await?;
    Ok(())
}

pub async fn remove_usage(
    db: &sqlx::PgPool,
    organisation_id: &str,
    resource_type: &str,
    resource_id: &str,
) -> AppResult<()> {
    sqlx::query(
        "DELETE FROM storage_usage WHERE organisation_id = $1 AND resource_type = $2 AND resource_id = $3",
    )
    .bind(organisation_id)
    .bind(resource_type)
    .bind(resource_id)
    .execute(db)
    .await?;
    Ok(())
}

pub async fn get_total_usage(
    db: &sqlx::PgPool,
    cache: &Cache,
    organisation_id: &str,
) -> AppResult<i64> {
    let cache_key = format!("storage:{}", organisation_id);
    if let Some(cached) = cache.get(&cache_key).await {
        if let Ok(val) = cached.parse::<i64>() {
            return Ok(val);
        }
    }

    let total: Option<i64> = sqlx::query_scalar(
        "SELECT SUM(size_bytes) FROM storage_usage WHERE organisation_id = $1",
    )
    .bind(organisation_id)
    .fetch_one(db)
    .await?;

    let total = total.unwrap_or(0);
    cache.set(&cache_key, &total.to_string(), 300).await;
    Ok(total)
}

pub async fn get_usage_breakdown(
    db: &sqlx::PgPool,
    organisation_id: &str,
) -> AppResult<Vec<StorageBreakdown>> {
    let rows: Vec<(String, i64, i64)> = sqlx::query_as(
        "SELECT resource_type, COALESCE(SUM(size_bytes), 0)::bigint, COUNT(*)::bigint FROM storage_usage WHERE organisation_id = $1 GROUP BY resource_type",
    )
    .bind(organisation_id)
    .fetch_all(db)
    .await?;

    Ok(rows
        .into_iter()
        .map(|(rt, total, count)| StorageBreakdown {
            resource_type: rt,
            total_bytes: total,
            count,
        })
        .collect())
}

pub async fn check_quota(
    db: &sqlx::PgPool,
    cache: &Cache,
    organisation_id: &str,
    additional_bytes: i64,
    limit_mb: i64,
) -> AppResult<bool> {
    if limit_mb < 0 {
        return Ok(true); // Unlimited
    }
    let current = get_total_usage(db, cache, organisation_id).await?;
    let limit_bytes = limit_mb * 1024 * 1024;
    Ok(current + additional_bytes <= limit_bytes)
}
